File: Array/medium/3338-count-submatrices-with-top-left-element-and-sum-less-than-k/count-submatrices-with-top-left-element-and-sum-less-than-k.go
func countSubmatrices(grid [][]int, k int) int {
	ROWs := len(grid)
	COLS := len(grid[0])
	dp := make([]int, COLS)
	res := 0

	for i := 0; i < ROWs; i++ {
		prev := 0
		for j := 0; j < COLS; j++ {
			dp[j] += grid[i][j]
			prev += dp[j]
			if prev <= k {
				res++
			} else {
				break
			}
		}
	}
	return res
}

File: Array/medium/3338-count-submatrices-with-top-left-element-and-sum-less-than-k/count-submatrices-with-top-left-element-and-sum-less-than-k.rs
impl Solution {
    pub fn count_submatrices(grid: Vec<Vec<i32>>, k: i32) -> i32 {
        let n = grid.len();
        let m = grid[0].len();
        let mut dp = vec! [0; m];
        let mut res = 0;

        for i in 0..n{
            let mut prev = 0;
            for j in 0..m{
                dp[j] += grid[i][j];
                prev += dp[j];
                if prev <= k{
                    res += 1
                }
            }
        }
        return res
    }
}